### PR TITLE
Added create volume feature using existing QOS

### PR DESCRIPTION
**Pull request by sonivi**
Implemented optional parameter 'qos-name', which is basically the name of 'vvset' present in 3par.
Created volume will be added in vvset which has QOS.

### CODE REVIEW

File: hpedockerplugin/hpe/hpe_3par_common.py [hunk 1]
```diff
@@ -939,18 +942,29 @@ def _set_flash_cache_policy_in_vvs(self, flash_cache, vvs_name):
...
-            self.client.deleteVolumeSet(vvs_name)
-            raise exception.PluginException(ex)
+                                  cpg, flash_cache, vvs_name=None):
+        if vvs_name is not None:
+            try:
+                if flash_cache is not None:
+                    self._set_flash_cache_policy_in_vvs(flash_cache, vvs_name)
+                self.client.addVolumeToVolumeSet(vvs_name, volume_name)
+            except Exception as ex:
+                msg = _('VV Set %s does not exist.') % vvs_name
```
Comment (imran-ansari) [inline hpedockerplugin/hpe/hpe_3par_common.py]: We can probably get exception for different reasons here in which case logging the actual reason would be more helpful.
E.g. msg = "Failed to set flash-cache policy or add volume to VV set %s - %s" % (vvs_name, ex.message)

File: hpedockerplugin/hpe_storage_api.py [hunk 2]
```diff
@@ -450,7 +450,7 @@ def volumedriver_create(self, name, opts=None):
         valid_volume_create_opts = ['mount-volume', 'compression',
                                     'size', 'provisioning', 'flash-cache',
                                     'cloneOf', 'snapshotOf', 'expirationHours',
-                                    'retentionHours', 'promote']
+                                    'retentionHours', 'promote', 'qos-name']
```
Comment (imran-ansari) [inline hpedockerplugin/hpe_storage_api.py]: Do we need to make qos-name mutually exclusive for cloneOf and snapshotOf?

Comment (wdurairaj) [inline hpedockerplugin/hpe_storage_api.py]: some validation you can add for presence of only -o flash-cache=true and -o qos-name=<vv_set>

Comment (wdurairaj) [reply hpedockerplugin/hpe_storage_api.py]: or only -o qos-name=<vv_set> .. As Imran pointed out cloneOf, snapshotOf should'nt present in the CLI along with -o qos-name=<vv_set>

File: hpedockerplugin/hpe_storage_api.py [hunk 3]
```diff
@@ -463,50 +469,45 @@ def volumedriver_create(self, name, opts=None):
...
+                return self.revert_to_snapshot(name, opts)
+            elif ('promote' in contents['Opts']):
+                msg = (_('while reverting volume to snapshot status only '
+                         'valid option is promote=<vol_name>'))
+                LOG.error(msg)
+                return json.dumps({u"Err": six.text_type(msg)})
 
-        valid_compression_opts = ['true', 'false']
+            # mutually exclusive options check
+            mutually_exclusive_list = ['snapshotOf', 'cloneOf', 'qos-name']
```
Comment (imran-ansari) [inline hpedockerplugin/hpe_storage_api.py]: Please add promote to the mutually_exclusive_list as well

Comment (sonivi) [reply hpedockerplugin/hpe_storage_api.py]: Added 'promote' option.

File: hpedockerplugin/hpe_storage_api.py [hunk 4]
```diff
@@ -463,50 +469,46 @@ def volumedriver_create(self, name, opts=None):
...
-            vol_size = int(contents['Opts']['size'])
+            if ('flash-cache' in contents['Opts']):
+                vol_flash = str(contents['Opts']['flash-cache'])
 
-        vol_prov = volume.DEFAULT_PROV
-        if ('Opts' in contents and contents['Opts'] and
-                'provisioning' in contents['Opts']):
-            vol_prov = str(contents['Opts']['provisioning'])
+            if ('qos-name' in contents['Opts']):
+                vol_flash = str(contents['Opts']['qos-name'])
```
Comment (wdurairaj) [inline hpedockerplugin/hpe_storage_api.py]: should this not be vol_qos  ? instead of vol_flash

**Pull request merged**